File: solutions/zig/perfect-numbers/1/perfect_numbers.zig
const std = @import("std");

pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

/// Asserts that `n` is nonzero.
pub fn classify(n: u64) Classification {
    std.debug.assert(n > 0);

    if (n == 1) {
        return .deficient;
    }

    var sum: u64 = 1;
    var i: u64 = 2;
    while (i * i <= n) : (i += 1) {
        if (n % i == 0) {
            sum += i;
            if (i * i != n) {
                sum += n / i;
            }
        }
    }

    if (sum < n) {
        return .deficient;
    } else if (sum > n) {
        return .abundant;
    } else {
        return .perfect;
    }
}

File: solutions/javascript/perfect-numbers/1/perfect-numbers.js
/** @param {number} n */
export const classify = (n) => {
    if (n <= 0) {
        throw new Error('Classification is only possible for natural numbers.');
    }
    if (n === 1) return 'deficient';

    let sum = 1;
    for (let i = 2; i * i <= n; i++) {
        if (n % i === 0) {
            sum += i;
            if (i * i !== n) {
                sum += n / i;
            }
        }
    }

    if (sum < n) return 'deficient';
    if (sum > n) return 'abundant';
    return 'perfect';
};
